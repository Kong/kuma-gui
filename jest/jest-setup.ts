import '@testing-library/jest-dom/extend-expect'
import '@testing-library/jest-dom'

// Polyfills `window.fetch` for Jest because it runs in a Node environment where fetch isn’t available. It initially looked like this would change with Node.js 18, but that is not so.
import 'isomorphic-fetch'

import dotenv from 'dotenv'

import { replaceAttributesSnapshotSerializer } from './jest-replace-attribute-snapshot-serializer'
import { setupMockServer } from '../src/api/setupMockServer'

/**
 * Ensures the tests run with the project’s environment variables being available.
 */
dotenv.config()

/**
 * amcharts need SVGPathElement defined for our tests to work.
 *
 * See: https://github.com/amcharts/amcharts4/issues/1387
 */
Object.defineProperty(window, 'SVGPathElement', { value: class extends HTMLElement { } })

/**
 * Kongponents uses generated UUIDs for several attribute values.
 * This breaks the project’s snapshot tests since they’re based on fully-mounted components
 * which also includes those from external sources like Kongponents.
 *
 * In order to stabilize the tests which otherwise fail because the attribute values are different every run,
 * we use a custom snapshot serializer to replace those attribute values with one fixed value.
 */
expect.addSnapshotSerializer(replaceAttributesSnapshotSerializer([
  'id',
  'aria-describedby',
  'aria-labelledby',
  'aria-controls',
  'data-tableid',
]))

const server = setupMockServer(import.meta.env.VITE_KUMA_API_SERVER_URL)

// Establish API mocking before all tests.
beforeAll(() => server.listen())
// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers())
// Clean up after the tests are finished.
afterAll(() => server.close())

export { server }
