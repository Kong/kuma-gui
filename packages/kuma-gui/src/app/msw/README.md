---
section: Packages
---
# msw

A module for adding MSW to the application for Service Worker based mocking for
statically served environments. We use this for serving our mocks during PR
previews.

