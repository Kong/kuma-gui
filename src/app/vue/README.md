---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

