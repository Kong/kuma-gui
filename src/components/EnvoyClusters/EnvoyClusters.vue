<template>
  <StatusInfo
    :has-error="hasError"
    :is-loading="isLoading"
  >
    <KCard border-variant="noBorder">
      <template #body>
        <Prism
          id="clusters"
          language="json"
          :code="clusters"
        />
      </template>
      <template #actions>
        <KClipboardProvider
          v-if="clusters"
          v-slot="{ copyToClipboard }"
        >
          <KPop placement="bottom">
            <KButton
              appearance="primary"
              @click="() => { copyToClipboard(clusters) }"
            >
              Copy config to clipboard
            </KButton>
            <template #content>
              <div>
                <p>Config copied to clipboard!</p>
              </div>
            </template>
          </KPop>
        </KClipboardProvider>
      </template>
    </kcard>
  </StatusInfo>
</template>

<script>
import Prism from 'vue-prismjs'

import Kuma from '@/services/kuma'
import StatusInfo from '@/components/Utils/StatusInfo'

export default {
  name: 'EnvoyClusters',
  components: {
    StatusInfo,
    Prism,
  },
  props: {
    mesh: {
      type: String,
      default: '',
    },
    dppName: {
      type: String,
      default: '',
    },
    zoneIngressName: {
      type: String,
      default: '',
    },
    zoneEgressName: {
      type: String,
      default: '',
    },
  },

  data() {
    return {
      isLoading: true,
      hasError: false,
    }
  },

  watch: {
    dppName() {
      this.fetchClusters()
    },
    zoneIngressName() {
      this.fetchClusters()
    },
    zoneEgressName() {
      this.fetchClusters()
    },
  },
  mounted() {
    this.fetchClusters()
  },

  methods: {
    async fetchClusters() {
      this.hasError = false
      this.isLoading = true

      try {
        let clusters = ''

        if (this.mesh !== '' && this.dppName !== '') {
          clusters = await Kuma.getDataplaneClusters({
            mesh: this.mesh,
            dppName: this.dppName,
          })
        } else if (this.zoneIngressName !== '') {
          clusters = await Kuma.getZoneIngressClusters({
            zoneIngressName: this.zoneIngressName,
          })
        } else if (this.zoneEgressName !== '') {
          clusters = await Kuma.getZoneEgressClusters({
            zoneEgressName: this.zoneEgressName,
          })
        }

        this.clusters = clusters
      } catch (e) {
        console.error(e)
        this.hasError = true
      } finally {
        this.isLoading = false
      }
    },
  },
}
</script>

<style lang="scss" scoped>
#clusters {
  max-height: 1000px;
}
</style>
