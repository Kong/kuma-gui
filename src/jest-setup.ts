import '@testing-library/jest-dom/extend-expect'
import '@testing-library/jest-dom'
import { replaceAttributesSnapshotSerializer } from './jest-replace-attribute-snapshot-serializer'
import { server as setupServer } from '@/services/mocks'

import './kongponents'

class SVGPathElement extends HTMLElement { }

// @ts-ignore
window.SVGPathElement = SVGPathElement

/**
 * Kongponents v6 uses a call to `crypto.getRandomValues` which isn’t implemented in jsdom by default.
 * To make the tests run, we supply it via Node’s `crypto` library.
 *
 * See: https://github.com/jsdom/jsdom/issues/1612
 */
Object.defineProperty(global.self, 'crypto', {
  value: {
    getRandomValues: (array: []) => require('crypto').randomBytes(array.length),
  },
})

/**
 * Kongponents v6 uses generated UUIDs for several attribute values.
 * This breaks the project’s snapshot tests since they’re based on fully-mounted components
 * which also includes those from external sources like Kongponents.
 *
 * In order to stabilize the tests which otherwise fail because the attribute values are different every run,
 * we use a custom snapshot serializer to replace those attribute values with one fixed value.
 */
expect.addSnapshotSerializer(replaceAttributesSnapshotSerializer([
  'id',
  'aria-describedby',
  'aria-labelledby',
  'aria-controls',
]))

const server = setupServer('http://localhost/')

// Establish API mocking before all tests.
beforeAll(() => server.listen())
// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers())
// Clean up after the tests are finished.
afterAll(() => server.close())

export { server }
