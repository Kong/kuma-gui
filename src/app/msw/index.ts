import { setupWorker } from 'msw/browser'

import { token } from '@/services/utils'
import type { Token, ServiceDefinition } from '@/services/utils'
import type { SetupWorker } from 'msw/browser'

type Handler = Parameters<typeof setupWorker>[0]
type Passthrough = (req: Request) => string | string[] | void

type Msw = ReturnType<SetupWorker['start']>

const $ = {
  msw: token<Msw>('msw'),
  mswHandlers: token('msw.handlers'),
  mswPassthroughs: token('msw.passthroughs'),
}
export const TOKENS = $
const passthroughExtension = /\.(vue|ts|js|json)(\?.*)?$/
export const services = (_: Record<string, Token>): ServiceDefinition[] => {
  return [
    [token('msw.passthrough.msw'), {
      service: () => [(req: Request) => {
        const { pathname, href } = new URL(req.url)
        if (
          !pathname.startsWith('/node_modules') &&
         !pathname.startsWith('/src/assets/') &&
         !passthroughExtension.test(href)
        ) {
          return `Found an unhandled ${req.method} request to ${href}`
        }
      }],
      labels: [
        $.mswPassthroughs,
      ],
    },
    ],
    [$.msw, {
      service: (handlers: Handler[] = [], passthroughs: Passthrough[]) => {
        return setupWorker(...handlers).start({
          quiet: true,
          onUnhandledRequest(req: Request) {
            passthroughs.reduce<string[]>((prev, item) => prev.concat(item(req) ?? []), []).forEach(item => {
              console.warn(item)
            })
          },
        })
      },
      arguments: [
        $.mswHandlers,
        $.mswPassthroughs,
      ],
    }],
  ]
}
