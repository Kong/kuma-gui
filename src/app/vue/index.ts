import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
