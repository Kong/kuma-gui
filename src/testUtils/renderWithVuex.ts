import { render, RenderOptions, RenderResult } from '@testing-library/vue'
import merge from 'lodash/merge'
import { StoreOptions } from 'vuex'
import Vue, { ComponentOptions } from 'vue'
import Store from '@/store'

function renderWithVuex(
  Component: typeof Vue | ComponentOptions<Vue>,
  options: RenderOptions<Vue> = { store: {}, routes: [] },
): RenderResult {
  const store: StoreOptions<any> = Store()

  // Render the component and merge the original store and the custom one
  // provided as a parameter. This way, we can alter some behaviors of the
  // initial implementation.

  return render(Component, { ...options, store: merge(store, options.store), routes: options.routes })
}

export default renderWithVuex
