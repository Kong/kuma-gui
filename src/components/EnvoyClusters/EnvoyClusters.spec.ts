import renderWithVuex from '@/testUtils/renderWithVuex';
import { screen } from '@testing-library/vue'
import { rest } from 'msw'
import { server } from '@/jest-setup'
import EnvoyClusters from './EnvoyClusters.vue'

describe('EnvoyClusters.vue', () => {
  it('renders snapshot', async () => {
    const { container } = renderWithVuex(EnvoyClusters, {
      props: {
        mesh: 'foo',
        dppName: 'dataplane-test-456',
      },
      routes: [],
    })

    expect(container).toMatchSnapshot()
  })

  it('renders loading', () => {
    renderWithVuex(EnvoyClusters, {
      props: {
        mesh: 'foo',
        dppName: 'dataplane-test-456',
      },
      routes: [],
    })

    expect(screen.getByText('spinner')).toBeInTheDocument()
  })

  it('renders error', async () => {
    jest.spyOn(console, 'error').mockImplementation(() => {}); // silence console errors

    server.use(
      rest.get('http://localhost/meshes/:mesh/dataplanes/:dataplaneName/clusters', (req, res, ctx) =>
        res(ctx.status(500), ctx.json({})),
      ),
    )

    renderWithVuex(EnvoyClusters, {
      props: {
        mesh: 'default',
        dppName: 'dataplane-test-456',
      },
      routes: [],
    })

    expect(await screen.findByText(/An error has occurred while trying to load this data./)).toBeInTheDocument()
  })
})
