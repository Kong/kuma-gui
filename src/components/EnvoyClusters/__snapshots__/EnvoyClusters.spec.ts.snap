// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnvoyClusters.vue renders snapshot 1`] = `
<div>
  <div>
    <!---->
     
    <div
      class="empty-state-wrapper"
      data-v-567e03f2=""
    >
      <div
        class="empty-state-title"
        data-v-567e03f2=""
      >
        <!---->
        <h5
          data-v-567e03f2=""
        >
          <div
            class="card-icon mb-3"
            data-v-567e03f2=""
          >
            <svg
              class="kong-icon kong-icon-spinner"
              data-v-567e03f2=""
              data-v-a7c9c1d2=""
              height="42"
              role="img"
              viewBox="0 0 16 16"
              width="42"
            >
              <title
                data-v-a7c9c1d2=""
              >
                spinner
              </title>
              <g
                data-v-a7c9c1d2=""
              >
                <path
                  d="M13.3154759 4.60336552c.6394482-.63889655.6394482-1.67503449 0-2.31393104-.6388966-.63889655-1.6744828-.63889655-2.3133793 0-.6394483.63889655-.6394483 1.67503449 0 2.31393104.6388965.63889655 1.6744827.63889655 2.3133793 0"
                  data-v-a7c9c1d2=""
                  fill="rgba(0, 0, 0, 0.1)"
                  opacity=".1"
                />
                <path
                  d="M13.9609931 6.17246897c-.9037241 0-1.6364138.73268965-1.6364138 1.63641379 0 .90317241.7326897 1.63586207 1.6364138 1.63586207.9037241 0 1.6358621-.73268966 1.6358621-1.63586207 0-.90372414-.732138-1.63641379-1.6358621-1.63641379"
                  data-v-a7c9c1d2=""
                  fill="rgba(0, 0, 0, 0.1)"
                  opacity=".25"
                />
                <path
                  d="M10.9938207 11.0105379c-.6388966.6388966-.6388966 1.6750345 0 2.3139311.6388965.6388965 1.6750345.6388965 2.313931 0 .6388966-.6388966.6388966-1.6750345 0-2.3139311-.6388965-.6388965-1.6750345-.6388965-2.313931 0"
                  data-v-a7c9c1d2=""
                  fill="rgba(0, 0, 0, 0.1)"
                  opacity=".35"
                />
                <path
                  d="M7.79988966 12.3247448c-.90372414 0-1.63586207.7326897-1.63586207 1.6364138 0 .9031724.73213793 1.6358621 1.63586207 1.6358621.90372413 0 1.63641379-.7326897 1.63641379-1.6358621 0-.9037241-.73268966-1.6364138-1.63641379-1.6364138"
                  data-v-a7c9c1d2=""
                  fill="rgba(0, 0, 0, 0.1)"
                  opacity=".5"
                />
                <path
                  d="M2.28706207 11.004469c-.63889655.6388965-.63889655 1.6750344 0 2.313931.63889655.6388966 1.67503448.6388966 2.31393103 0 .63889656-.6388966.63889656-1.6750345 0-2.313931-.63889655-.6388966-1.67503448-.6388966-2.31393103 0"
                  data-v-a7c9c1d2=""
                  fill="rgba(0, 0, 0, 0.1)"
                  opacity=".65"
                />
                <path
                  d="M3.27244138 7.80888276c0-.90372414-.73268966-1.63641379-1.63641379-1.63641379-.90372414 0-1.63586207.73268965-1.63586207 1.63641379 0 .90317241.73213793 1.63586207 1.63586207 1.63586207.90372413 0 1.63641379-.73268966 1.63641379-1.63586207"
                  data-v-a7c9c1d2=""
                  fill="rgba(0, 0, 0, 0.1)"
                  opacity=".75"
                />
                <path
                  d="M4.59271724 4.60943448c.63889655-.63889655.63889655-1.67503448 0-2.31393103-.63889655-.63889655-1.67503448-.63889655-2.31393103 0-.63889655.63889655-.63889655 1.67503448 0 2.31393103.63889655.63889655 1.67503448.63889655 2.31393103 0"
                  data-v-a7c9c1d2=""
                  fill="rgba(0, 0, 0, 0.1)"
                  opacity=".85"
                />
                <path
                  d="M6.16402759 1.6361931c0 .90372414.73213793 1.6364138 1.63586207 1.6364138.90372413 0 1.63641379-.73268966 1.63641379-1.6364138 0-.90372413-.73268966-1.63641379-1.63641379-1.63641379-.90372414 0-1.63586207.73268966-1.63586207 1.63641379z"
                  data-v-a7c9c1d2=""
                  fill="rgba(0, 0, 0, 0.1)"
                  opacity=".95"
                />
              </g>
            </svg>
          </div>
          
      Data Loading...
    
        </h5>
      </div>
      <div
        class="empty-state-content"
        data-v-567e03f2=""
      >
        <p
          data-v-567e03f2=""
        />
        <p
          data-v-567e03f2=""
        >
          <!---->
        </p>
      </div>
    </div>
     
    <!---->
     
    <!---->
  </div>
</div>
`;
